block-producer/src/state_view/mod.rs

```diff
@@ -0,0 +1,217 @@
+    ) -> Result<(), ApplyBlockError>;
+}
+
+pub struct DefaulStateView<TI, Store>
+where
+    TI: GetTxInputs,
+    Store: ApplyBlock,
+{
+    get_tx_inputs: Arc<TI>,
+
+    store: Arc<Store>,
+
+    /// The account ID of accounts being modified by transactions currently in the block production
+    /// pipeline. We currently ensure that only 1 tx/block modifies any given account.
+    accounts_in_flight: Arc<RwLock<BTreeSet<AccountId>>>,
+
+    /// The nullifiers of notes consumed by transactions currently in the block production pipeline.
+    nullifiers_in_flight: Arc<RwLock<BTreeSet<Digest>>>,
+}
+
+impl<TI, Store> DefaulStateView<TI, Store>
+where
+    TI: GetTxInputs,
+    Store: ApplyBlock,
+{
+    pub fn new(
+        get_tx_inputs: Arc<TI>,
+        store: Arc<Store>,
+    ) -> Self {
+        Self {
+            get_tx_inputs,
+            store,
+            accounts_in_flight: Arc::new(RwLock::new(BTreeSet::new())),
+            nullifiers_in_flight: Arc::new(RwLock::new(BTreeSet::new())),
+        }
+    }
+}
+
+#[async_trait]
+impl<TI, Store> TransactionVerifier for DefaulStateView<TI, Store>
+where
+    TI: GetTxInputs,
+    Store: ApplyBlock,
+{
+    // TODO: Verify proof as well
+    async fn verify_tx(
+        &self,
+        candidate_tx: SharedProvenTx,
+    ) -> Result<(), VerifyTxError> {
+        // 1. soft-check if `tx` violates in-flight requirements.
+        //
+        // This is a "soft" check, because we'll need to redo it at the end. We do this soft check
+        // to quickly reject clearly infracting transactions before hitting the store (slow).
+        ensure_in_flight_constraints(
+            candidate_tx.clone(),
+            &*self.accounts_in_flight.read().await,
+            &*self.nullifiers_in_flight.read().await,
+        )?;
+
+        // 2. Fetch the transaction inputs from the store, and check tx input constraints
+        let tx_inputs = self.get_tx_inputs.get_tx_inputs(candidate_tx.clone()).await?;
+        ensure_tx_inputs_constraints(candidate_tx.clone(), tx_inputs)?;
+
+        // 3. Re-check in-flight transaction constraints, and if verification passes, register
+        //    transaction
+        //
+        // Note: We need to re-check these constraints because we dropped the locks since we last
+        // checked
+        {
+            let mut locked_accounts_in_flight = self.accounts_in_flight.write().await;
+            let mut locked_nullifiers_in_flight = self.nullifiers_in_flight.write().await;
+
+            ensure_in_flight_constraints(
+                candidate_tx.clone(),
+                &locked_accounts_in_flight,
+                &locked_nullifiers_in_flight,
+            )?;
+
+            // Success! Register transaction as successfully verified
+            locked_accounts_in_flight.insert(candidate_tx.account_id());
+
+            let mut nullifiers_in_tx: BTreeSet<_> =
+                candidate_tx.consumed_notes().iter().map(|note| note.nullifier()).collect();
```

Comment:
I wonder if we should add a helper method on `ProvenTransaction` - something like `nullifiers()` which would return an iterator over all nullifies created by the transaction.